File: ios/Memorize/Network/TextParser.swift
import Foundation

class TextParser {
    static func parseJson(_ jsonString: String) -> ParsedTextStructure? {
        // Clean JSON from markdown code blocks if present
        let cleanedJson = jsonString
            .replacingOccurrences(of: "```json", with: "")
            .replacingOccurrences(of: "```", with: "")
            .trimmingCharacters(in: .whitespacesAndNewlines)
        
        guard let data = cleanedJson.data(using: .utf8) else {
            return nil
        }
        
        do {
            let decoder = JSONDecoder()
            let structure = try decoder.decode(ParsedTextStructure.self, from: data)
            return structure
        } catch {
            print("Error parsing JSON: \(error)")
            return nil
        }
    }
}

File: android/app/src/main/java/com/memorize/network/TextParser.kt
package com.memorize.network

import com.google.gson.Gson
import com.google.gson.JsonObject
import com.google.gson.JsonParser

object TextParser {
    private val gson = Gson()
    
    fun parseJson(jsonString: String): ParsedTextStructure? {
        return try {
            // Try to extract JSON from markdown code blocks if present
            val cleanedJson = jsonString
                .replace("```json", "")
                .replace("```", "")
                .trim()
            
            val jsonObject = JsonParser().parse(cleanedJson).asJsonObject
            val sectionsArray = jsonObject.getAsJsonArray("sections")
            
            val sections = sectionsArray.map { sectionElement ->
                val sectionObj = sectionElement.asJsonObject
                val paragraphsArray = sectionObj.getAsJsonArray("paragraphs")
                
                val paragraphs = paragraphsArray.map { paragraphElement ->
                    val paragraphObj = paragraphElement.asJsonObject
                    val phrasesArray = paragraphObj.getAsJsonArray("phrases")
                    
                    val phrases = phrasesArray.map { phraseElement -> phraseElement.asString }
                    ParagraphStructure(phrases)
                }
                
                SectionStructure(paragraphs)
            }
            
            ParsedTextStructure(sections)
        } catch (e: Exception) {
            e.printStackTrace()
            null
        }
    }
}

